formats/html.ts

```diff
@@ -21,24 +73,340 @@ export class HtmlImporter extends FormatImporter {
+					if (typeof ast === "string") {
+						return { text: ast };
+					}
+					try {
+						let { link, link: { path: linkpath, display }, text } = ast;
+						const correctedPath = linkpath.startsWith("//") ? `https:${linkpath}` : linkpath;
+						let url;
+						try {
+							url = new URL(correctedPath);
+						} catch (e) {
+							if (!(e instanceof TypeError)) {
+								throw e;
+							}
+							url = new URL(normalizePath(correctedPath)
+								.split("/")
+								.map(encodeURIComponent)
+								.join("/"), pathURL);
+						}
+						const attachment = await this.downloadAttachmentCached(mdFile, url);
+						if (attachment instanceof TFile) {
+							text = this.app.fileManager.generateMarkdownLink(attachment, mdFile.path, "", display);
+						}
+						return { text, attachment, link } as const;
+					} catch (e) {
+						console.error(e);
+						return { text: ast.text };
+					}
+				}));
+			results.total += transformedAST
+				.filter(({ link }) => link)
+				.length;
+			results.failed = results.failed.concat(transformedAST
+				.filter(({ attachment }) => attachment === "failed")
+				.map(({ link: { path } }) => path));
+			results.skipped = results.skipped.concat(transformedAST
+				.filter(({ attachment }) => attachment === "skipped")
+				.map(({ link: { path } }) => path));
+
+			mdContent = transformedAST.map(({ text }) => text).join("");
+			await this.app.vault.modify(mdFile, mdContent);
+		} catch (e) {
+			console.error(e);
+			results.failed.push(file.toString());
+			if (mdFile) {
+				try {
+					await this.app.vault.delete(mdFile);
+				} catch (e) {
+					console.error(e);
+				}
+			}
+		}
+		return results;
+	}
+
+	downloadAttachmentCached(mdFile: TFile, url: URL) {
+		return this.attachments[url.href] ??= this.downloadAttachment(mdFile, url);
+	}
+
+	async downloadAttachment(mdFile: TFile, url: URL) {
+		try {
+			let response;
+			switch (url.protocol) {
+				case "file:":
+					response = await this.requestFile(url);
+					break;
+				case "https:":
+				case "http:":
+					response = await this.requestHTTP(url);
+					break;
+				default:
+					throw new Error(url.href);
+			}
+			if (!await this.filterAttachment(response)) {
+				return "skipped";
+			}
+			let filename = getURLFilename(url);
+			const { data, mime: actualMime } = response;
+			if ((mime(splitFilename(filename).extension) || "application/octet-stream") !== actualMime) {
+				const ext = extension(actualMime);
+				if (ext) {
+					filename += `.${ext}`;
+				}
+			}
+			return await this.writeAttachment(mdFile, filename, data);
+		} catch (e) {
+			console.error(e);
+			return "failed";
+		}
+	}
+
+	async requestFile(url: URL) {
+		const data = (await fsPromises.readFile(nodeUrl.fileURLToPath(url.href))).buffer;
+		return { mime: await detectMime(url, data), data };
+	}
+
+	async requestHTTP(url: URL) {
+		url = new URL(url.href);
+		let response;
+		try {
+			url.protocol = "https:";
+			response = await requestUrl({
```

Comment:
Can we try to use `fetch` first before falling back to requestUrl? I believe it's much more efficient since it doesn't have to go through a cross-process call through the electron main process.

Reply:
Done.